### PR TITLE
Pause mining during JoinSplit creation

### LINKED ISSUE #1522: Pause mining while doing a JoinSplit
JoinSplit latency will be terrible if the node is simultaneously mining. By default, when a JoinSplit starts we should cancel all mining threads' Equihash solvers (to free memory) and keep those threads paused until the JoinSplit operation queue is empty again. This won't interfere with dedicated mining processes because they won't be doing JoinSplits, and it won't interfere with nodes that are just being used for wallet operations, but it is a good default for nodes doing "casual" mining.

### PR DESCRIPTION
**Pull request by aniemerg**
Closes #1522 .

### CODE REVIEW

**Comment by ebfull** [comment]
Awesome!

Is there no way to disable mining via the RPC? (To avoid the scenario where this code would *re-enable* mining that has been disabled in some other way.)

**Comment by aniemerg** [comment]
So, as best as I understand it, mining is enabled/disabled in the rpc via setgenerate(). If we were to use setgenerate(), we'd have to first check whether mining is enabled with getgenerate(), store that in a flag, disable mining, and then after the joinsplit, re-enable mining if appropriate. I can certainly take that approach (and submit a PR to that effect). 

I avoided doing that, because it seemed slightly more complex, but also because this approach avoids changing mapArgs, as it would under the above rpc approach. Since mapArgs appears intended to reflect the parameter state set by the user, I wanted to avoid writing code that would change it. Suppose, for example, the user starts a z_sendmany, and then sends an rpc command to disable mining. Under the RPC approach I described above, it's possible that upon the completion of the joinsplit that mining is re-enabled against the wishes of the user.

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 1]
```diff
@@ -144,6 +145,10 @@ bool AsyncRPCOperation_sendmany::main_impl() {
 
     assert(isfromtaddr_ != isfromzaddr_);
 
+#ifdef ENABLE_WALLET
```
Comment (str4d) [inline src/wallet/asyncrpcoperation_sendmany.cpp]: If #1795 is merged before this PR, change this `#ifdef` and the one below to:

```
#if defined(ENABLE_WALLET) && defined(ENABLE_MINING)
```

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 2]
```diff
@@ -144,6 +145,10 @@ bool AsyncRPCOperation_sendmany::main_impl() {
...
+    GenerateBitcoins(false, NULL, 0);
+#endif
```
Comment (str4d) [inline src/wallet/asyncrpcoperation_sendmany.cpp]: If #1965 is merged before this PR, change this entire `#ifdef` block to:

```
#ifdef ENABLE_MINING
  #ifdef ENABLE_WALLET
    GenerateBitcoins(false, NULL, 0);
  #else
    GenerateBitcoins(false, 0);
  #endif
#endif
```

and likewise below.

**Comment by aniemerg** [comment]
Should address desired changes. #1795 and #1965 should be merged first.

**Comment by ebfull** [comment]
Thank you @aniemerg! ACK

@zkbot r+

**Comment by ebfull** [comment]
@zkbot r-

**Comment by ebfull** [comment]
(Merging #1795 and #1965 first.)

**Comment by ebfull** [comment]
Given the time requirements for 1.0.5 I will need to bump this to 1.0.6.

**Comment by bitcartel** [comment]
Waiting for #1863 and #1965 to be merged first.

**Comment by bitcartel** [comment]
We should have a test for this.

**Comment by aniemerg** [comment]
Regarding a test, I've been unsure how to accomplish that using the rpc testing framework. The way that this PR currently works is to turn off mining in a way that isn't revealed when using rpc commands such as getgenerate. So I am not sure how to write the most natural test -- that is, turn on mining, start a joinsplit, and  check to see that mining is off.

**Comment by daira** [comment]
Re @bitcartel's point 3, I don't think it's a problem if `zcbenchmark createjoinsplit` doesn't pause mining. That RPC is intended to be used with a node (and machine, since it measures wall clock time) that is otherwise completely idle; just pausing mining on an active node would not be sufficient to get an accurate result.

**Comment by arcalinea** [comment]
@aniemerg if we can fix @bitcartel's first point, then we'll merge and open another ticket to address point 4. 
If we can fix the first point by mid-week, we'll include this in 1.0.7.

**Comment by aniemerg** [comment]
@arcalinea commit 54c92a6 was intended to fix @bitcartel 's first point. I believe it still needs review.

**Comment by arcalinea** [comment]
Squashed the commits. Can we get one more ACK before merging though? (@str4d?)

**Comment by arcalinea** [comment]
Thanks. Opened ticket for point 4, merging

@zkbot r+

**Pull request merged**